File: src/infrastructure/analyzers/device_boundary_analyzer.py
"""Device boundary (bezel) detector for phone/tablet screen attacks.

Ported from biometric-processor's DeviceBoundaryDetector.
Detects rectangular frames (phone/tablet bezels) surrounding faces
using Canny edge detection + Hough line segments + contour analysis.

This is the most direct physical signal for screen-based attacks:
if there's a phone bezel around the face, it's a screen replay.
"""

import time
from collections import deque
from typing import Optional

import cv2
import numpy as np

from src.domain.models import FaceROI, AnalyzerResult

# Common phone/tablet aspect ratios
DEVICE_ASPECT_RATIOS = (16.0 / 9.0, 19.5 / 9.0, 18.0 / 9.0, 4.0 / 3.0)


def _clamp01(v: float) -> float:
    return max(0.0, min(1.0, float(v)))


class DeviceBoundaryAnalyzer:
    """Detects phone/tablet screen bezels around tracked faces.

    Searches for rectangular contours and straight line segments
    in a padded region around each face. Validates against common
    device aspect ratios (16:9, 19.5:9, 18:9, 4:3).

    This analyzer needs the FULL FRAME (not just face crop) because
    it looks for bezel edges AROUND the face. Use set_frame() before
    analyze(), or pass via the pipeline's frame context.
    """

    def __init__(self, padding_ratio: float = 0.55, spoof_threshold: float = 0.50):
        self._padding = max(0.05, padding_ratio)
        self._spoof_threshold = spoof_threshold
        self._original_frame: np.ndarray | None = None

    @property
    def name(self) -> str:
        return "device_boundary"

    def set_frame(self, frame: np.ndarray):
        self._original_frame = frame

    def analyze(self, face_crop: np.ndarray, face_roi: FaceROI) -> AnalyzerResult:
        start = time.perf_counter()

        if self._original_frame is None:
            return AnalyzerResult(name=self.name, score=50.0,
                                  details={"error": "no_frame"}, elapsed_ms=0)

        frame = self._original_frame
        bbox = face_roi.bbox
        fh, fw = frame.shape[:2]

        # Extract padded ROI around face
        pad_x = int(bbox.width * self._padding)
        pad_y = int(bbox.height * self._padding)
        rx1 = max(0, bbox.x1 - pad_x)
        ry1 = max(0, bbox.y1 - pad_y)
        rx2 = min(fw, bbox.x2 + pad_x)
        ry2 = min(fh, bbox.y2 + pad_y)
        roi = frame[ry1:ry2, rx1:rx2]

        if roi.size == 0:
            return AnalyzerResult(name=self.name, score=100.0,
                                  details={"no_roi": True},
                                  elapsed_ms=(time.perf_counter() - start) * 1000)

        roi_h, roi_w = roi.shape[:2]

        # Edge detection
        gray = cv2.cvtColor(roi, cv2.COLOR_BGR2GRAY)
        blurred = cv2.GaussianBlur(gray, (5, 5), 0)
        edges = cv2.Canny(blurred, 45, 140)

        # Hough lines
        min_line = max(18, int(min(roi_h, roi_w) * 0.18))
        lines = cv2.HoughLinesP(edges, 1.0, np.pi / 180, 36,
                                 minLineLength=min_line,
                                 maxLineGap=max(10, int(min(roi_h, roi_w) * 0.05)))

        line_score = self._analyze_lines(lines, roi_h, roi_w)
        contour_score = self._analyze_contours(edges, roi_h, roi_w, bbox, rx1, ry1)

        # Combined boundary score (0-1, higher = more bezel-like)
        boundary_score = 0.50 * contour_score + 0.50 * line_score

        # Convert to liveness score (0-100, higher = more live-like)
        # High boundary score = spoof, so invert
        if boundary_score >= self._spoof_threshold:
            score = max(0.0, 30.0 * (1.0 - boundary_score))  # Strong spoof signal
        else:
            score = 70.0 + 30.0 * (1.0 - boundary_score / self._spoof_threshold)

        elapsed_ms = (time.perf_counter() - start) * 1000

        return AnalyzerResult(
            name=self.name,
            score=max(0.0, min(100.0, score)),
            details={
                "boundary_score": boundary_score,
                "line_score": line_score,
                "contour_score": contour_score,
                "bezel_detected": boundary_score >= self._spoof_threshold,
                "n_lines": len(lines) if lines is not None else 0,
            },
            elapsed_ms=elapsed_ms,
        )

    def _analyze_lines(self, lines: Optional[np.ndarray], roi_h: int, roi_w: int) -> float:
        if lines is None or len(lines) == 0:
            return 0.0

        horizontal = []
        vertical = []
        total_length = 0.0

        for raw in lines[:, 0]:
            x1, y1, x2, y2 = [float(v) for v in raw]
            length = float(np.hypot(x2 - x1, y2 - y1))
            if length < max(18.0, min(roi_w, roi_h) * 0.18):
                continue
            angle = abs(np.degrees(np.arctan2(y2 - y1, x2 - x1))) % 180.0
            total_length += length

            if min(angle, abs(angle - 180.0)) <= 18.0:
                horizontal.append(length)
            elif abs(angle - 90.0) <= 18.0:
                vertical.append(length)

        parallel = min(1.0, 0.5 * min(len(horizontal), 2) + 0.5 * min(len(vertical), 2))
        orthogonal = 1.0 if horizontal and vertical else 0.0
        density = min(1.0, total_length / max(float((roi_w + roi_h) * 2), 1.0))

        return min(1.0, 0.45 * parallel + 0.35 * orthogonal + 0.20 * density)

    def _analyze_contours(self, edges: np.ndarray, roi_h: int, roi_w: int,
                          face_bbox, roi_x: int, roi_y: int) -> float:
        contours, _ = cv2.findContours(edges, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        roi_area = float(roi_h * roi_w)
        face_cx = face_bbox.x1 + face_bbox.width / 2
        face_cy = face_bbox.y1 + face_bbox.height / 2

        best_score = 0.0
        for contour in contours:
            area = float(cv2.contourArea(contour))
            if area < roi_area * 0.04:
                continue
            perimeter = cv2.arcLength(contour, True)
            if perimeter <= 1.0:
                continue
            approx = cv2.approxPolyDP(contour, 0.03 * perimeter, True)
            if len(approx) != 4:
                continue

            x, y, w, h = cv2.boundingRect(approx)
            abs_x, abs_y = roi_x + x, roi_y + y

            # Must be bigger than face
            if w < face_bbox.width * 1.1 or h < face_bbox.height * 1.1:
                continue
            # Face center must be inside rectangle
            if not (abs_x <= face_cx <= abs_x + w and abs_y <= face_cy <= abs_y + h):
                continue

            rectangularity = area / max(float(w * h), 1.0)
            aspect = max(float(w / max(h, 1)), float(h / max(w, 1)))
            aspect_score = max(0.0, 1.0 - min(abs(aspect - t) for t in DEVICE_ASPECT_RATIOS) / 0.55)
            face_cover = min(1.0, min(w / max(float(face_bbox.width), 1.0),
                                       h / max(float(face_bbox.height), 1.0)) / 2.1)
            area_ratio = float(w * h) / max(roi_area, 1.0)

            score = min(1.0,
                        0.35 * _clamp01((rectangularity - 0.60) / 0.35)
                        + 0.30 * aspect_score
                        + 0.20 * face_cover
                        + 0.15 * _clamp01((area_ratio - 0.20) / 0.45))

            best_score = max(best_score, score)

        return best_score

File: web/src/infrastructure/analyzers/DeviceBoundaryAnalyzer.ts
// Port of src/infrastructure/analyzers/device_boundary_analyzer.py:1-189
//
// Device Boundary — fusion weight 2.5.
//
// Detects phone/tablet bezels surrounding a tracked face. The Python
// source uses OpenCV (cv2.Canny + cv2.HoughLinesP + cv2.findContours +
// cv2.approxPolyDP). The browser equivalent could lazy-load OpenCV.js
// (~10 MB), but the task spec asks for the lighter path:
//
//   * Canvas2D + native Sobel-derived edge map.
//   * Hand-rolled probabilistic Hough line-segment scan.
//   * Skip the contour rectangle search (cv2.findContours +
//     approxPolyDP) — replaced with an axis-aligned line-merge that
//     reconstructs candidate device rectangles from horizontal/vertical
//     line clusters. This is a documented ~10% accuracy hit but keeps
//     the bundle dep-free.
//
// All line- and contour-side scoring weights are preserved verbatim so
// the calibration of `boundary_score >= 0.50 → spoof` still applies.

import {
  AnalyzerResult,
  FaceROI,
  IFaceAnalyzer,
  makeAnalyzerResult,
} from "../../domain/models";
import { SourceImage, toImageData } from "../../utils/imageOps";

// Common phone/tablet aspect ratios — verbatim from Python:21.
const DEVICE_ASPECT_RATIOS = [16 / 9, 19.5 / 9, 18 / 9, 4 / 3] as const;

export interface DeviceBoundaryOptions {
  /** Padding ratio around the face bbox. Default 0.55. */
  paddingRatio?: number;
  /** Boundary score above which we declare bezel-detected. Default 0.50. */
  spoofThreshold?: number;
  /**
   * If true, downsample the ROI to MAX_DIM=160 before processing.
   * Cuts work by ~16x at the cost of edge precision; acceptable for the
   * "did we find a phone-sized rectangle" question. Default true.
   */
  downsample?: boolean;
}

const DOWNSAMPLE_MAX = 160;

export class DeviceBoundaryAnalyzer implements IFaceAnalyzer {
  readonly name = "device_boundary";

  private readonly paddingRatio: number;
  private readonly spoofThreshold: number;
  private readonly downsample: boolean;
  private currentFrame: SourceImage | null = null;

  constructor(options: DeviceBoundaryOptions = {}) {
    this.paddingRatio = Math.max(0.05, options.paddingRatio ?? 0.55);
    this.spoofThreshold = options.spoofThreshold ?? 0.50;
    this.downsample = options.downsample !== false;
  }

  /** Set the current full frame (analyzer needs surroundings of the face). */
  setFrame(frame: SourceImage): void {
    this.currentFrame = frame;
  }

  analyze(_faceCrop: ImageData | null, face: FaceROI): AnalyzerResult {
    const start = performance.now();

    if (!this.currentFrame) {
      return makeAnalyzerResult(
        this.name,
        50.0,
        { error: "no_frame" },
        performance.now() - start,
      );
    }

    const frame = toImageData(this.currentFrame);
    const fw = frame.width;
    const fh = frame.height;
    const bbox = face.bbox;

    const padX = Math.floor(bbox.width * this.paddingRatio);
    const padY = Math.floor(bbox.height * this.paddingRatio);
    const rx1 = Math.max(0, bbox.x1 - padX);
    const ry1 = Math.max(0, bbox.y1 - padY);
    const rx2 = Math.min(fw, bbox.x2 + padX);
    const ry2 = Math.min(fh, bbox.y2 + padY);
    const roiW = rx2 - rx1;
    const roiH = ry2 - ry1;

    if (roiW <= 8 || roiH <= 8) {
      return makeAnalyzerResult(
        this.name,
        100.0,
        { no_roi: true },
        performance.now() - start,
      );
    }

    // === Build a downsampled grayscale buffer of the ROI. ===
    let workW = roiW;
    let workH = roiH;
    let scale = 1;
    if (this.downsample) {
      scale = Math.max(1, Math.ceil(Math.max(roiW, roiH) / DOWNSAMPLE_MAX));
      workW = Math.floor(roiW / scale);
      workH = Math.floor(roiH / scale);
    }
    const gray = new Float32Array(workW * workH);
    const stride = fw * 4;
    for (let y = 0; y < workH; y++) {
      const sy = ry1 + Math.min(roiH - 1, y * scale);
      let srcOff = sy * stride + rx1 * 4;
      for (let x = 0; x < workW; x++) {
        const off = srcOff + Math.min(roiW - 1, x * scale) * 4;
        // luminance ≈ 0.299R + 0.587G + 0.114B
        gray[y * workW + x] =
          0.299 * frame.data[off] +
          0.587 * frame.data[off + 1] +
          0.114 * frame.data[off + 2];
      }
      srcOff += stride;
    }

    // === Sobel edge magnitudes, then threshold to a binary edge map. ===
    const edges = sobelBinary(gray, workW, workH);

    // === Hand-rolled line scan: detect strong horizontal/vertical streaks. ===
    const minLine = Math.max(8, Math.floor(Math.min(workH, workW) * 0.18));
    const lines = scanLines(edges, workW, workH, minLine);

    const lineScore = scoreLines(lines, workH, workW);

    // === Rectangle reconstruction from line clusters (replaces cv2 contours). ===
    const contourScore = scoreReconstructedRectangles(
      lines,
      workW,
      workH,
      bbox,
      rx1,
      ry1,
      scale,
    );

    const boundaryScore = 0.5 * contourScore + 0.5 * lineScore;

    let score: number;
    if (boundaryScore >= this.spoofThreshold) {
      score = Math.max(0.0, 30.0 * (1.0 - boundaryScore));
    } else {
      score = 70.0 + 30.0 * (1.0 - boundaryScore / this.spoofThreshold);
    }
    score = Math.max(0.0, Math.min(100.0, score));

    return makeAnalyzerResult(
      this.name,
      score,
      {
        boundary_score: round(boundaryScore, 4),
        line_score: round(lineScore, 4),
        contour_score: round(contourScore, 4),
        bezel_detected: boundaryScore >= this.spoofThreshold,
        n_lines: lines.length,
      },
      performance.now() - start,
    );
  }

  reset(): void {
    this.currentFrame = null;
  }
}

/** Sobel + magnitude threshold → binary edge map (Uint8Array, 0 or 255). */
function sobelBinary(
  gray: Float32Array,
  w: number,
  h: number,
): Uint8Array {
  const out = new Uint8Array(w * h);
  let maxMag = 0;
  const mag = new Float32Array(w * h);
  for (let y = 1; y < h - 1; y++) {
    for (let x = 1; x < w - 1; x++) {
      const i = y * w + x;
      const gx =
        -gray[i - w - 1] -
        2 * gray[i - 1] -
        gray[i + w - 1] +
        gray[i - w + 1] +
        2 * gray[i + 1] +
        gray[i + w + 1];
      const gy =
        -gray[i - w - 1] -
        2 * gray[i - w] -
        gray[i - w + 1] +
        gray[i + w - 1] +
        2 * gray[i + w] +
        gray[i + w + 1];
      const m = Math.sqrt(gx * gx + gy * gy);
      mag[i] = m;
      if (m > maxMag) maxMag = m;
    }
  }
  // Threshold at ~Otsu-lite: 0.35 * max. Calibrated against Canny(45,140).
  // Guard against the uniform-image degeneracy: with maxMag≈0, a 0.35*max
  // threshold of 0 would mark every pixel as an edge. Require some absolute
  // gradient before any pixel can be on.
  if (maxMag < 1.0) return out;
  const thresh = 0.35 * maxMag;
  for (let i = 0; i < mag.length; i++) {
    out[i] = mag[i] >= thresh ? 255 : 0;
  }
  return out;
}

interface DetectedLine {
  /** "h" (horizontal streak) or "v" (vertical streak). */
  kind: "h" | "v";
  /** For "h": y-row. For "v": x-col. */
  major: number;
  /** Start of the run along the perpendicular axis. */
  start: number;
  /** End of the run (inclusive). */
  end: number;
  /** Length in pixels (end - start + 1). */
  length: number;
}

/**
 * Scan the binary edge map for horizontal and vertical lines using row/col
 * runs. This is the "probabilistic Hough" approximation:
 *   * For each row, collapse the binary into a 1D series and pick contiguous
 *     edge-pixel runs of length >= minLine.
 *   * Same for each column.
 * Diagonal lines are intentionally ignored — bezels are axis-aligned in
 * a typical capture geometry.
 */
function scanLines(
  edges: Uint8Array,
  w: number,
  h: number,
  minLine: number,
): DetectedLine[] {
  const out: DetectedLine[] = [];
  const minRun = Math.max(2, minLine);

  // Horizontal scan.
  for (let y = 0; y < h; y++) {
    let runStart = -1;
    let runDensity = 0;
    let runLen = 0;
    for (let x = 0; x <= w; x++) {
      const on = x < w && edges[y * w + x] !== 0;
      if (on) {
        if (runStart < 0) runStart = x;
        runDensity += 1;
        runLen += 1;
      } else if (runStart >= 0) {
        // Allow up to 5% gap pixels per row by extending into close gaps.
        const gapTolerance = 2;
        if (
          x < w &&
          edges[y * w + Math.min(w - 1, x + gapTolerance)] !== 0
        ) {
          runLen += 1;
          continue;
        }
        if (runLen >= minRun && runDensity / runLen >= 0.7) {
          out.push({
            kind: "h",
            major: y,
            start: runStart,
            end: runStart + runLen - 1,
            length: runLen,
          });
        }
        runStart = -1;
        runDensity = 0;
        runLen = 0;
      }
    }
  }

  // Vertical scan.
  for (let x = 0; x < w; x++) {
    let runStart = -1;
    let runDensity = 0;
    let runLen = 0;
    for (let y = 0; y <= h; y++) {
      const on = y < h && edges[y * w + x] !== 0;
      if (on) {
        if (runStart < 0) runStart = y;
        runDensity += 1;
        runLen += 1;
      } else if (runStart >= 0) {
        if (runLen >= minRun && runDensity / runLen >= 0.7) {
          out.push({
            kind: "v",
            major: x,
            start: runStart,
            end: runStart + runLen - 1,
            length: runLen,
          });
        }
        runStart = -1;
        runDensity = 0;
        runLen = 0;
      }
    }
  }

  return out;
}

/** Mirrors `_analyze_lines()` in the Python source. */
function scoreLines(
  lines: DetectedLine[],
  roiH: number,
  roiW: number,
): number {
  if (lines.length === 0) return 0.0;
  const minLen = Math.max(18.0, Math.min(roiW, roiH) * 0.18);
  const horizontal: DetectedLine[] = [];
  const vertical: DetectedLine[] = [];
  let totalLength = 0;
  for (const ln of lines) {
    if (ln.length < minLen) continue;
    totalLength += ln.length;
    if (ln.kind === "h") horizontal.push(ln);
    else vertical.push(ln);
  }
  const parallel = Math.min(
    1.0,
    0.5 * Math.min(horizontal.length, 2) + 0.5 * Math.min(vertical.length, 2),
  );
  const orthogonal = horizontal.length > 0 && vertical.length > 0 ? 1.0 : 0.0;
  const density = Math.min(
    1.0,
    totalLength / Math.max((roiW + roiH) * 2, 1),
  );
  return Math.min(1.0, 0.45 * parallel + 0.35 * orthogonal + 0.2 * density);
}

/**
 * Approximation of the Python `_analyze_contours()`:
 * Pair the strongest horizontal and vertical lines into candidate
 * rectangles; score each one by aspect-ratio match to a known device
 * format and by face-center inclusion. ~10% accuracy hit vs the OpenCV
 * version per the file header.
 *
 * Coordinates are in WORK space (downsampled ROI). We rescale into
 * absolute frame space using `roiX0`, `roiY0`, `scale`.
 */
function scoreReconstructedRectangles(
  lines: DetectedLine[],
  workW: number,
  workH: number,
  faceBbox: { width: number; height: number; x1: number; y1: number; x2: number; y2: number },
  roiX0: number,
  roiY0: number,
  scale: number,
): number {
  const horizontal = lines.filter((l) => l.kind === "h");
  const vertical = lines.filter((l) => l.kind === "v");
  if (horizontal.length < 2 || vertical.length < 2) return 0;

  // Sort by length desc; consider top 6 of each direction (cap combinatorics at 6×6×6×6=1296).
  horizontal.sort((a, b) => b.length - a.length);
  vertical.sort((a, b) => b.length - a.length);
  const hCand = horizontal.slice(0, 6);
  const vCand = vertical.slice(0, 6);

  const faceCx = faceBbox.x1 + faceBbox.width / 2;
  const faceCy = faceBbox.y1 + faceBbox.height / 2;
  const roiArea = workW * workH;

  let best = 0;
  for (let i = 0; i < hCand.length; i++) {
    for (let j = i + 1; j < hCand.length; j++) {
      const top = hCand[i].major < hCand[j].major ? hCand[i] : hCand[j];
      const bot = hCand[i].major < hCand[j].major ? hCand[j] : hCand[i];
      const yTop = top.major;
      const yBot = bot.major;
      const rh = yBot - yTop;
      if (rh < 8) continue;

      for (let m = 0; m < vCand.length; m++) {
        for (let k = m + 1; k < vCand.length; k++) {
          const left =
            vCand[m].major < vCand[k].major ? vCand[m] : vCand[k];
          const right =
            vCand[m].major < vCand[k].major ? vCand[k] : vCand[m];
          const xLeft = left.major;
          const xRight = right.major;
          const rw = xRight - xLeft;
          if (rw < 8) continue;

          // Closure: each corner pair must overlap.
          if (top.start > xRight || top.end < xLeft) continue;
          if (bot.start > xRight || bot.end < xLeft) continue;
          if (left.start > yBot || left.end < yTop) continue;
          if (right.start > yBot || right.end < yTop) continue;

          // === Translate work-space rect into absolute frame space. ===
          const absX = roiX0 + xLeft * scale;
          const absY = roiY0 + yTop * scale;
          const absW = rw * scale;
          const absH = rh * scale;

          // Must be bigger than the face.
          if (absW < faceBbox.width * 1.1 || absH < faceBbox.height * 1.1)
            continue;
          // Face center must be inside the rectangle.
          if (
            !(
              absX <= faceCx &&
              faceCx <= absX + absW &&
              absY <= faceCy &&
              faceCy <= absY + absH
            )
          )
            continue;

          const rectArea = rw * rh;
          // "Rectangularity": how much of the bounding rect is filled by
          // the inferred contour. A 4-line boundary covers the perimeter
          // pixels — approximate fill ratio from edge density of the four
          // segments vs the rectangle perimeter.
          const perim = 2 * (rw + rh);
          const edgePixels = top.length + bot.length + left.length + right.length;
          const rectangularity = clamp01(edgePixels / Math.max(perim, 1));

          const aspect = Math.max(rw / Math.max(rh, 1), rh / Math.max(rw, 1));
          let aspectErr = Infinity;
          for (const t of DEVICE_ASPECT_RATIOS) {
            const e = Math.abs(aspect - t);
            if (e < aspectErr) aspectErr = e;
          }
          const aspectScore = Math.max(0, 1 - aspectErr / 0.55);

          const faceCover = Math.min(
            1.0,
            Math.min(
              absW / Math.max(faceBbox.width, 1),
              absH / Math.max(faceBbox.height, 1),
            ) / 2.1,
          );
          const areaRatio = rectArea / Math.max(roiArea, 1);

          const score = Math.min(
            1.0,
            0.35 * clamp01((rectangularity - 0.6) / 0.35) +
              0.30 * aspectScore +
              0.20 * faceCover +
              0.15 * clamp01((areaRatio - 0.2) / 0.45),
          );
          if (score > best) best = score;
        }
      }
    }
  }
  return best;
}

function clamp01(v: number): number {
  return Math.max(0, Math.min(1, v));
}

function round(x: number, digits: number): number {
  const k = Math.pow(10, digits);
  return Math.round(x * k) / k;
}
